### PR TITLE
TagsInput: Design update and component refactor

### PR DESCRIPTION
**Pull request by torkelo**
Not happy with the TagsInput design and how it looks in forms

* Rewrite to FC and simpify it (no needs to internal tags state)
* Tried tags as input prefix, it looks better but Input component does not handle dynamic width prefixes
* Make Add button a link & suffix style (inside input)
* Improve placeholder text
* Link button disabled when no tag name specified 

Before:
Screenshot from 2021-02-12 11-40-44

After:
Screenshot from 2021-02-12 11-33-36

### CODE REVIEW

File: packages/grafana-ui/src/components/TagsInput/TagsInput.tsx [hunk 1]
```diff
@@ -1,122 +1,88 @@
...
-  constructor(props: Props) {
-    super(props);
-
-    this.state = {
-      newTag: '',
-      tags: this.props.tags || [],
-    };
-  }
+export const TagsInput: FC<Props> = ({ placeholder = 'New tag (enter key to add)', tags = [], onChange }) => {
+  const [newTagName, setNewName] = useState<string>('');
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/TagsInput/TagsInput.tsx:16]: I think the types are inferred from default value, so need to specify them. 
```suggestion
  const [newTagName, setNewName] = useState('');
```

File: packages/grafana-ui/src/components/TagsInput/TagsInput.tsx [hunk 2]
```diff
@@ -1,122 +1,88 @@
...
   };
 
-  // Using React.MouseEvent to avoid tslint error
-  onAdd = (event: React.MouseEvent) => {
+  const onAdd = (event: React.MouseEvent) => {
     event.preventDefault();
-    if (this.state.newTag !== '') {
-      this.setNewTags();
-    }
+    onChange(tags.concat([newTagName]));
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/TagsInput/TagsInput.tsx:29]:
```suggestion
    onChange(tags.concat(newTagName));
```
Don't think this needs to be an array.

File: packages/grafana-ui/src/components/TagsInput/TagsInput.tsx [hunk 3]
```diff
@@ -1,122 +1,88 @@
...
-          ...prevState,
-          tags: [...prevState.tags, prevState.newTag],
-          newTag: '',
-        }),
-        () => this.onChange()
-      );
-    } else {
-      this.setState({ newTag: '' });
+    if (event.key === 'Enter' && newTagName !== '') {
+      onChange(tags.concat([newTagName]));
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/TagsInput/TagsInput.tsx:36]:
```suggestion
      onChange(tags.concat(newTagName));
```

File: packages/grafana-ui/src/components/TagsInput/TagsInput.tsx [hunk 4]
```diff
@@ -1,122 +1,88 @@
...
-          {tags &&
-            tags.map((tag: string, index: number) => {
-              return <TagItem key={`${tag}-${index}`} name={tag} onRemove={this.onRemove} />;
-            })}
-        </div>
+  return (
+    <div className={styles.wrapper}>
+      <div className={styles.tags}>
+        {tags &&
+          tags.map((tag: string, index: number) => {
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/TagsInput/TagsInput.tsx:45]:
```suggestion
        {tags?.map((tag: string, index: number) => {
```

**Review by Clarity-89 — APPROVED** [review]
Not a big fan of the left tag positioning, I think displaying them below the input is better. Also it doesn't prevent from adding duplicate tags (not sure if it should). Otherwise looks good!

**Comment by torkelo** [comment]
@Clarity-89 thanks for the review!

> Not a big fan of the left tag positioning

think it's much better than below, below feels like it's part of something else, disconnected from the input. I would prefer if it's inside the input but the input component does not support dynamic prefixes, have to update how prefixes are done in input to support that

**Pull request merged**